add CLIPlugin to accept arbitrary R shapes/JSON on the command-line

wihobbs opened the pull request:
This PR takes the excellent work done by @sam-maloney for [RFC#46](github.com/flux-framework/rfc/pull/465) and extends it into a command-line plugin that users can opt in to by setting `FLUX_CLI_PLUGINPATH`. This was [originally suggested by @grondo](https://github.com/flux-framework/rfc/pull/465#issuecomment-3201763842) as a way for us to test/get user feedback on the ergonomics of specifying shape on the command line, and make progress toward incorporating that option into our standard command line.

The plugin is relatively straightforward and the testing just references examples from the corresponding RFC. I know this is coming up right before many 🌴s, so no rush on reviews, but I wanted to get it to a stopping point before my own holiday begins :).

sam-maloney committed:
python: RFC 46 jobspec resource string parser

Problem: There is no parser for the CLI Jobspec Resource String Syntax
proposed in RFC 46.

Add a flux.shape.parser Python module which implements a parser
using the 'ply' module (mimicking the constraint parser).

For testing, the parser may be run from the command line with

 flux python -m flux.shape.parser [OPTIONS] -- EXPRESSION

wihobbs committed:
cli: add CLIPlugin to specify resources via shape or json

Problem: currently, users have no way on the CLI of utilizing the
experimental parser in RFC 46 for specifying a job shape or a resource
set they'd like to use via JSON. This causes users to defer to
cumbersome piping and JSON mangling when they want to submit jobs of
arbitrary resource shapes on the command line.

Add long options via plugin `--resources-json` & `--resources-shape`
for specifying resources on the command line beyond the standard ones.
Provide this plugin by default; package it with flux-core into
`flux.cli.plugins` and add it to the `default_plugins` in
src/bindings/python/flux/cli/plugin.py.

wihobbs committed:
t: add tests for jobspec shape cli plugin

Problem: the jobspec shape/JSON plugin is currently not tested on
the command line.

Add a sharness test.

grondo reviewed (commented):
Nice work! Just a few comments or questions to address inline. Thanks!

grondo commented on src/bindings/python/flux/Makefile.am:
```diff
@@ -139,8 +144,13 @@ constraint/parsetab.py: constraint/parser.py
+
 clean-local:
-	-rmdir constraint 2>/dev/null || :
+	-rmdir shape 2>/dev/null || :
```
`-rmdir constraint` was inadvertently removed here.

grondo commented on src/bindings/python/flux/shape/plugin/shape.py:
```diff
@@ -0,0 +1,59 @@
+from flux.shape.parser import ShapeParser
+
+
+class RFC44ShapePlugin(CLIPlugin):
```
Rename `RFC46ShapePlugin`

grondo commented on src/bindings/python/flux/shape/plugin/shape.py:
```diff
@@ -0,0 +1,59 @@
+        )
+
+    def preinit(self, args):
+        if getattr(args, "json") or getattr(args, "shape"):
```
`getattr(args, "json")` is equivalent to `args.json` but the latter is shorter and more idiomatic. I'd switch this to simply `if args.json or args.shape`

grondo commented on t/python/t0037-job-shape-parser.py:
```diff
@@ -0,0 +1,277 @@
+#!/usr/bin/env python3
+###############################################################
```
Actually this tests `ShapeParser` nothing to do with the CLI jobspec parser plugin.

grondo reviewed (commented):
It is getting to the point where @sam-maloney's `--resource-shape` option will be really useful so I made another pass on this one.

I'm also wondering if we should install the plugin so it is picked up by default, instead of requiring users to modify their plugin path to get the feature.

wihobbs committed:
t: add sharness tests for `flux python -m flux.shape.parser`

Problem: the RFC 46 resource shape parser is tested via its
corresponding CLIPlugin and with Python unit tests, but not as a
standalone tool.

Add t0039-shape-parser.t to fill the testing gap.

Assisted-By: Claude:claude-sonnet-4.5

wihobbs committed:
t: add python unit test for `ShapeParser`

Problem: there are examples of valid job shapes listed in RFC 46
but there is currently no unit test for the python jobspec shape
parser.

Add a unit test with the examples in RFC 46.

grondo commented on src/bindings/python/flux/shape/plugin/shape.py:
```diff
@@ -0,0 +1,59 @@
+                except TypeError:
+                    jobspec.jobspec["resources"] = data
+        else:
+            pass
```
Is the
```python
    else:
        pass
```
unnecessary here?

Also, just a suggestion: The function would be easier to read if it used `args.shape` and `args.json` directly instead of assigning to temporary local variables `s` and `j`.

wihobbs commented:
Ok, rebased again. The first two commits of this PR deal with outstanding issues in the CLIPlugin interface that would be required to merge. I'll go make a separate PR for those in a minute, since they're a separate issue from including a `--resource-shape` option.

This PR now includes:

- tests passing on el8 container and compatibility with the ancient pyyaml there (locally 🤞)
- `--resource-shape` and `--resource-json` included by default as residents of `/etc/cli/plugins` in the repo, but instaled to `$libexecdir/cli/plugins` on an installation
- and some mild shuffling of files to enable ^.

wihobbs commented:
Interestingly, `flux-sched`'s CI appears to be catching a legitimate error not caught by core:
```
  flux-submit: ERROR: 'types.SimpleNamespace' object has no attribute 'resources_shape'
  ```

The plugin may not be checking to make sure that a job without `--resources-shape` or `--resources-json` is ignored, or perhaps there's a subtlety I missed there.

sam-maloney commented on etc/cli/plugins/shape.py:
```diff
@@ -0,0 +1,55 @@
+        self.add_option(
+            "--shape",
+            metavar="SHAPE",
+            help="Provide an RFC 44 jobspec shape on the command line. Any other resource arguments will be ignored.",
```
```suggestion
            help="Provide an RFC 46 jobspec shape on the command line. Any other resource arguments will be ignored.",
```

wihobbs replied:
I've made that mistake a bunch. I don't know why in my head canon, the shape RFC is 44 🤦

sam-maloney commented on etc/cli/plugins/shape.py:
```diff
@@ -0,0 +1,55 @@
+            metavar="SHAPE",
+            help="Provide an RFC 44 jobspec shape on the command line. Any other resource arguments will be ignored.",
+        )
+        self.add_option(
+            "--json",
+            metavar="FILE",
+            help="Provide a JSON file specifying the `resources` section of a jobspec.",
+        )
```
Could be added later, but since RFCs 14 & 25 give all their examples as YAML it might make sense to also have an option for a YAML file as input.

wihobbs replied:
I think I'll defer this to a later PR, but good suggestion. (The reason is that our only yaml2json converter, albeit trivial would require writing more testing to get it through and I'd like to land this sooner rather than later)

sam-maloney commented on t/t0039-shape-parser.t:
```diff
@@ -0,0 +1,235 @@
+	cat >expected27 <<-EOF &&
+	[{"type": "node", "count": 1, "exclusive": true}]
+	EOF
+	$parser -- "node{x:true}" >output27 &&
```
```suggestion
	$parser -- "node{x}" >output27 &&
```
`true` should be implied if no value is given, and explicit flags are already tested with `node{key:true,flag:false}` a few test prior, so this should hopefully help with coverage to hit [L559](https://app.codecov.io/gh/flux-framework/flux-core/pull/7252?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=flux-framework#92375911843bba91d45474f27c792b19-R559)
(This is also probably the most idiomatic use case, as the whole point of this special flag is to keep everything as short as possible!)

wihobbs replied:
Bah! Silly Claude!

sam-maloney commented on t/t0039-shape-parser.t:
```diff
@@ -0,0 +1,235 @@
+	$parser --rangedict -- "slot=3+:2/node" >output32 &&
+	test_cmp expected32 output32
+'
+test_done
```
A couple of possible extra things to test for complete coverage:

- brackets around the count value, e.g. `node=[2]` to hit [L461](https://app.codecov.io/gh/flux-framework/flux-core/pull/7252?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=flux-framework#92375911843bba91d45474f27c792b19-R461)
- empty dict, e.g. `node{}` to hit [L552](https://app.codecov.io/gh/flux-framework/flux-core/pull/7252?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=flux-framework#92375911843bba91d45474f27c792b19-R522)
- slot with other dict items, e.g. `slot{task,foo:bar}/node` to hit [L425-L427](https://app.codecov.io/gh/flux-framework/flux-core/pull/7252?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=flux-framework#92375911843bba91d45474f27c792b19-R425)
- invalid token, e.g. `node=;` to hit [L345](https://app.codecov.io/gh/flux-framework/flux-core/pull/7252?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=flux-framework#92375911843bba91d45474f27c792b19-R345)
- unclosed quote, e.g. `'node` to hit [L118](https://app.codecov.io/gh/flux-framework/flux-core/pull/7252?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=flux-framework#92375911843bba91d45474f27c792b19-R118)
- illegal character, e.g. `node=a` to hit [L109](https://app.codecov.io/gh/flux-framework/flux-core/pull/7252?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=flux-framework#92375911843bba91d45474f27c792b19-R109)

The first two (count with brackets and empty dict) could simply be worked into existing test cases as they don't change the output in any way, but could also be added as separate cases to keep things as atomic as possible, whichever you prefer.

wihobbs replied:
These are excellent suggestions, thanks! I'll add all of them.

sam-maloney commented on src/bindings/python/flux/shape/parser.py:
```diff
@@ -0,0 +1,678 @@
+            return t
+        # If not empty, QSTRING token(s) were already returned,
+        # don't return the closing quote
+        return None
```
```suggestion
        else:
            # If not empty, QSTRING token(s) were already returned,
            # don't return the closing quote
            t = None
        return t
```
I'm not really sure why the security-bot is complaining, but perhaps this is more idiomatic? No strong preference either way on my end, but thanks for catching this edge case!

wihobbs replied:
I think the security bot is complaining about something else we changed ages ago. I like your suggested change better, will add.

sam-maloney reviewed (approved):
Thanks for all your work on this @wihobbs! Sorry I was on vacation last week and it took me a while to make a review, but it looks great to me, with only a handful of very minor comments :smile:

wihobbs committed:
t: edit t2717-python-cli-plugins.t test of --list-plugins

Problem: the --list-plugins test asserts that only 5 plugins should
be listed with that option. Until recently that was true, but as plugins
get added to the `flux.cli.plugins` namespace this number will go up.
In general, it seems cumbersome to edit that test each time a new
plugin is added to that namespace.

Amend the logic to check for just one known plugin in the output.

wihobbs committed:
t: edit python/t0019-cli-plugin.py conflict detection test

Problem: there is a test in python/t0019-cli-plugin.py that checks
that if two plugins that have conflicting options are loaded, the
first one in the FLUX_CLI_PLUGINPATH* wins. Or so it seemed! All this
test was really doing was checking that the registry didn't load both
plugins, and never actually checked that the right plugin was loaded.

Edit the test to fix this oversight.

wihobbs commented:
@grondo, do you want to take another look at this for a secondary approval/change request?

grondo commented on t/t2718-python-cli-job-shape.t:
```diff
@@ -0,0 +1,143 @@
+
+. $(dirname $0)/sharness.sh
+
+test_under_flux 4 job
```
Do these tests require a 4 node instance. If they do require a Flux instance, could you get away with 1 rank?

wihobbs replied:
I like keeping the first test around, because I think it's always good to test that there isn't some logic in the plugin that fails if you _don't_ use it, but you're right -- no need for a 4 node instance. Will swap to 1.

grondo commented:
I noticed one quick thing. Will circle back again later to check this out again.

garlick commented:
@wihobbs - I know you've been tasked with some testing due early next week, but at a lower priority than that, it would be great to get this PR finalized for our next tag.

wihobbs commented:
This is awaiting a review from @grondo who I believe is also juggling some high-priority items, but we agreed the same in our 1-1 this week. I expect we'll make the May release cut off. ;)

grondo commented:
Was just taking a look at this. @wihobbs I guess you get the honor of fixing one last merge conflict in `t/Makefile.am`

grondo commented on src/bindings/python/flux/cli/plugins/shape.py:
```diff
@@ -0,0 +1,61 @@
+                jobspec.jobspec["resources"] = ShapeParser().parse(args.shape)
+            elif args.json:
+                with open(args.json, "r") as json_file:
+                    data = json.loads(json_file.read())
```
Could use `json.load(json_file)` here.

grondo commented on src/bindings/python/flux/cli/plugins/shape.py:
```diff
@@ -0,0 +1,61 @@
+                        jobspec.jobspec["resources"] = data["resources"]
+                    except TypeError:
+                        jobspec.jobspec["resources"] = data
+        except AttributeError:
```
This `except` catches `AttributeError` from anywhere in this entire block, including an error in `ShapeParser()` or `json.loads` and swallows it. The `try/except` block could be more targeted here, or maybe more simply just put an early exit at the top:
```python
if not (getattr(args, "json", None) or getattr(args, "shape", None)):
      return
# Now one of args.json or args.shape is non-None
```

grondo reviewed (approved):
LGTM! A couple minor things inline.

Going forward, I wonder if we should remove the default_plugins list and instead just load all plugins found in the `flux.cli.plugin` namespace, since that's what is done with other plugin loading schemes in Flux and (I think?) with FLUX_CLI_PLUGINPATH.

wihobbs commented:
Thanks @grondo! Setting MWP.

wihobbs commented:
>I guess you get the honor of fixing one last merge conflict in t/Makefile.am

Spoke too soon. There was yet another conflict. In the same file, on the same line.

garlick commented:
Yay!
